nixos/modules/services/web-servers/nginx/default.nix

```diff
@@ -639,8 +638,18 @@ in
         ''
         ${cfg.preStart}
         ${cfg.package}/bin/nginx -c ${configFile} -p ${cfg.stateDir} -t
+
+        # The nginx configuration test might have created log files as root,
+        # make sure the nginx user/group can write to them.
+        chown -R ${cfg.user}:${cfg.group} ${cfg.stateDir}
```

Comment:
There is already a chown in the default cfg.preStart -- and yes, it's a bit ugly that it needs to be done before *and* after, but I couldn't convince the nginx binary to not create the log files as root:root when in test mode. I toyed with running the nginx config test unprivileged, but unfortunately this then failed because the config test fails to bind 80/443.

Reply:
I use this method:

```
  services.logrotate = {
    config = ''
      ${cfgNginxSpool}/logs/access.log
      ${cfgNginxSpool}/logs/error.log
      {
        create 0640 ${cfg.user} ${cfg.group}
        postrotate
          [ ! -f /run/nginx.pid ] || kill -USR1 `cat /run/nginx.pid`
        endscript
      }
    '';
  };
```

Reply:
Please check thats variant:

```
      preStart =  mkOption {
        type = types.lines;
        default = ''
          test -d "${cfg.stateDir}" || mkdir -m 750 "${cfg.stateDir}"
          test -d "${cfg.stateDir}/logs" || mkdir -m 750 -p "${cfg.stateDir}/logs"
          test `stat -c %a:%U:%G "${cfg.stateDir}"` = "750:${cfg.user}:${cfg.group}" || (chmod 750 "${cfg.stateDir}" && chown ${cfg.user}:${cfg.group} "${cfg.stateDir}")
          test `stat -c %a:%U:%G "${cfg.stateDir}/logs"` = "750:${cfg.user}:${cfg.group}" || (chmod 750 "${cfg.stateDir}/logs" && chown ${cfg.user}:${cfg.group} "${cfg.stateDir}/logs")
          ${pkgs.su-exec}/bin/su-exec ${cfg.user}:${cfg.group} /run/wrappers/bin/nginx -c ${configFile} -p ${cfg.stateDir} -t
        '';
...
    systemd.services.nginx = {
      preStart = 
      ''
        ${cfg.preStart}
      '';
...
    security.wrappers.nginx = {
      source = "${pkgs.nginx}/bin/nginx";
      capabilities = "cap_net_bind_service+eip";
      owner = cfg.user;
      group = cfg.group;
      permissions = "u+rx,g+x";
    };

```

Reply:
Please update preStart to:
 ```
         test -d "${cfg.stateDir}" || mkdir -m 750 "${cfg.stateDir}"
          test -d "${cfg.stateDir}/logs" || mkdir -m 750 -p "${cfg.stateDir}/logs"
          test `stat -c %a:%U:%G "${cfg.stateDir}"` = "750:${cfg.user}:${cfg.group}" || (chmod 750 "${cfg.stateDir}" && chown ${cfg.user}:${cfg.group} "${cfg.stateDir}")
          test `stat -c %a:%U:%G "${cfg.stateDir}/logs"` = "750:${cfg.user}:${cfg.group}" || (chmod 750 "${cfg.stateDir}/logs" && chown ${cfg.user}:${cfg.group} "${cfg.stateDir}/logs")
```
Thanks.

Reply:
I don't understand why the setuid binary is required here?

Reply:
I would rather prefer starting nginx as root then having a setuid binary, where every member of the `nginx` group could become start nginx as the nginx user.

Reply:
It was a mistake to expose `preStart` as an option in https://github.com/NixOS/nixpkgs/pull/48875 in the first place. Now we have to do ugly workarounds.

Reply:
My prestart scripts normal worked, not ugly. 
nginx -t running from root incorrectly sets permissions log

Reply:
@Mic92 please make sure you're looking at the latest version of this PR -- it has close to no changes to the preStart script. This comment thread was started on an older version.

@Izorkin I don't understand why the preStart script permissions dance should change in any way from what is currently in master. The current version in this PR does the right thing.

Reply:
@delroth i have no deprecation in PR.